Unable to join tables if there are more than 1000 rows

Code to reproduce:
```
import numpy as np
import modin.pandas as pd
import modin.config as cfg

cfg.IsExperimental.put(True)
cfg.Engine.put("Native")
cfg.StorageFormat.put("Hdk")

data = np.random.randint(0, 100, size=(1001, 99))
df = pd.DataFrame(data)
df["new"] = pd.Series(list(range(len(df))))
print(df)
```

Error message:
```
modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:76: in _executeRelAlgJson
    res = rel_alg_executor.execute()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   RuntimeError: Hash join failed, reason(s): Could not build a 1-to-1 correspondence for columns involved in equijoin | Cannot join on rowid | Cannot fall back to loop join for non-trivial inner table size

_sql.pyx:103: RuntimeError
```

The following cases are probably also related to this issue. In all these cases inner join by index is performed.

```
data = {"A": []}
df = pd.DataFrame(data)
df["B"] = pd.Series([1, 2, 3])
assert len(df) == 3
```
E       assert 0 == 3

```
data = {"a": [1, 2, 3, 4]}
index = pandas.to_datetime(["2020-02-06", "2020-02-06", "2020-02-22", "2020-03-26"])
df = pd.DataFrame(data, index=index)
df["b"] = pd.Series(np.arange(4))
print(df)
```
E   RuntimeError: Cannot compare between TIMESTAMP[ns] and INT64

--- a/omniscidb/QueryEngine/GroupByRuntime.cpp
+++ b/omniscidb/QueryEngine/GroupByRuntime.cpp
@@ -284,6 +284,14 @@ bucketized_hash_join_idx(int64_t hash_buff,
 }
 
 extern "C" RUNTIME_EXPORT ALWAYS_INLINE DEVICE int64_t
+rowid_hash_join_idx(const int64_t key, const int64_t min_key, const int64_t max_key) {
+  if (key >= min_key && key <= max_key) {
+    return key;
+  }
+  return -1;
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE DEVICE int64_t
 hash_join_idx(int64_t hash_buff,
               const int64_t key,
               const int64_t min_key,
@@ -331,6 +339,14 @@ bucketized_hash_join_idx_bitwise(int64_t hash_buff,
                                                     min_key,
                                                     translated_val,
                                                     bucket_normalization);
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE DEVICE int64_t
+rowid_hash_join_idx_nullable(const int64_t key,
+                             const int64_t min_key,
+                             const int64_t max_key,
+                             const int64_t null_val) {
+  return key != null_val ? rowid_hash_join_idx(key, min_key, max_key) : -1;
 }
 
 extern "C" RUNTIME_EXPORT ALWAYS_INLINE DEVICE int64_t

--- a/omniscidb/QueryEngine/JoinHashTable/PerfectJoinHashTable.cpp
+++ b/omniscidb/QueryEngine/JoinHashTable/PerfectJoinHashTable.cpp
@@ -434,7 +434,19 @@ ColumnsForDevice PerfectJoinHashTable::fetchColumnsForDevice(
   for (const auto& inner_outer_pair : inner_outer_pairs_) {
     const auto inner_col = inner_outer_pair.first;
     if (inner_col->isVirtual()) {
-      throw FailedToJoinOnVirtualColumn();
+      if (inner_outer_pairs_.size() != 1) {
+        throw HashJoinFail("Only single predicate row-id joins are currently supported.");
+      }
+      // push back an empty join column, which will prevent hash table build. the probe
+      // for a row-id hash join simply does null checks and returns the input row-id
+      auto type = inner_col->type();
+      join_columns.emplace_back(
+          JoinColumn{/*ptr=*/nullptr,
+                     /*col_chunks_buff_sz=*/0,
+                     /*num_chunks=*/1,
+                     /*num_elems=*/0,
+                     /*elem_sz=*/static_cast<size_t>(type->size())});
+      continue;
     }
     join_columns.emplace_back(fetchJoinColumn(inner_col,
                                               fragments,
@@ -971,11 +983,44 @@ llvm::Value* PerfectJoinHashTable::codegenSlot(const CompilationOptions& co,
   }
   const auto key_lvs = code_generator.codegen(key_col, true, co);
   CHECK_EQ(size_t(1), key_lvs.size());
+  auto key_col_type = key_col->type();
+
+  if (val_col->isVirtual()) {
+    // short-circuit hash table load. we know the key will be the row number we want to
+    // access, so just return the key
+    CHECK(val_col_var);  // make sure this virtual column is a column
+
+    std::vector<llvm::Value*> hash_join_idx_args{
+        executor_->cgen_state_->castToTypeIn(key_lvs.front(), 64)};
+
+    const auto& query_info = getInnerQueryInfo(val_col_var).info;
+    if (query_info.getPhysicalNumTuples() == 0) {
+      // empty inner table, set range to empty / invalid
+      hash_join_idx_args.push_back(executor_->cgen_state_->llInt(/*min=*/int64_t(0)));
+      hash_join_idx_args.push_back(executor_->cgen_state_->llInt(/*max=*/int64_t(-1)));
+    } else {
+      hash_join_idx_args.push_back(
+          executor_->cgen_state_->llInt(rhs_source_col_range_.getIntMin()));
+      hash_join_idx_args.push_back(
+          executor_->cgen_state_->llInt(rhs_source_col_range_.getIntMax()));
+    }
+
+    auto key_col_logical_type = key_col->type()->canonicalize();
+    if (key_col_logical_type->nullable() || isBitwiseEq()) {
+      hash_join_idx_args.push_back(executor_->cgen_state_->llInt(
+          inline_fixed_encoding_null_value(key_col_logical_type)));
+    }
+
+    std::string fname{"rowid_hash_join_idx"};
+    if (key_col_type->nullable()) {
+      fname += "_nullable";
+    }
+    return executor_->cgen_state_->emitCall(fname, hash_join_idx_args);
+  }
   auto hash_ptr = codegenHashTableLoad(index);
   CHECK(hash_ptr);
   const auto hash_join_idx_args = getHashJoinArgs(hash_ptr, key_col, co);
 
-  auto key_col_type = key_col->type();
   std::string fname(key_col_type->isDate() ? "bucketized_hash_join_idx"s
                                            : "hash_join_idx"s);
 

--- a/omniscidb/Tests/ArrowBasedExecuteTest.cpp
+++ b/omniscidb/Tests/ArrowBasedExecuteTest.cpp
@@ -17849,6 +17849,58 @@ TEST_F(Select, ParseIntegerExceptions) {
   }
 }
 
+TEST_F(Select, RowIdJoin) {
+  // make sure select tests pass under trivial join loop threshold default
+  EXPECT_NO_THROW(run_multiple_agg(
+      R"(SELECT t1.x, t2.y FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.rowid AND t1.x = t2.y;)",
+      ExecutorDeviceType::CPU,
+      /*allow_loop_join=*/false));
+
+  // disable loop joins, force hash join path
+  const auto trivial_join_loop_state = config().exec.join.trivial_loop_join_threshold;
+  ScopeGuard reset = [&] {
+    config().exec.join.trivial_loop_join_threshold = trivial_join_loop_state;
+  };
+  config().exec.join.trivial_loop_join_threshold = 1;
+
+  for (auto dt : {ExecutorDeviceType::CPU, ExecutorDeviceType::GPU}) {
+    SKIP_NO_GPU();
+
+    c(R"(SELECT t1.x, t2.y FROM test t1 LEFT JOIN test_inner t2 ON t1.rowid = t2.rowid;)",
+      dt);
+    c(R"(SELECT t1.x, t2.y, COUNT(*) FROM test t1 LEFT JOIN test_inner t2 ON t1.rowid = t2.rowid GROUP BY 1, 2 ORDER BY 1, 2 NULLS LAST;)",
+      dt);
+
+    c(R"(SELECT t1.x, t2.y FROM test t1 LEFT JOIN test_empty t2 ON t1.rowid = t2.rowid;)",
+      dt);
+
+    c(R"(SELECT t1.x, t2.y FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.rowid;)",
+      dt);
+    c(R"(SELECT t1.x, t2.y, COUNT(*) FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.rowid GROUP BY 1, 2 ORDER BY 1, 2;)",
+      dt);
+    c(R"(SELECT t1.x, t2.y FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.rowid WHERE t1.x > 7;)",
+      dt);
+    c(R"(SELECT t1.x, t2.y FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.rowid WHERE t2.x > 0;)",
+      dt);
+
+    EXPECT_EQ(
+        int64_t(1),
+        v<int64_t>(run_simple_agg(
+            R"(SELECT COUNT(*) FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.x;)",
+            dt)));
+    EXPECT_EQ(
+        int64_t(0),
+        v<int64_t>(run_simple_agg(
+            R"(SELECT COUNT(*) FROM test t1 INNER JOIN test_inner t2 ON t1.x = t2.rowid;)",
+            dt)));
+
+    EXPECT_ANY_THROW(run_multiple_agg(
+        R"(SELECT t1.x, t2.y FROM test t1 INNER JOIN test_inner t2 ON t1.rowid = t2.rowid AND t1.x = t2.y;)",
+        dt,
+        /*allow_loop_join=*/false));
+  }
+}
+
 class SubqueryTestEnv : public ExecuteTestBase, public ::testing::Test {
  protected:
   void SetUp() override {

